### PR TITLE
Sinusoid example in paper.md

### PR DESCRIPTION
**Pull request by odunbar**
## Purpose 

Adds the sinusoid example #228 into the paper

## Content

Latest pdf: 
[paper.pdf](https://github.com/CliMA/EnsembleKalmanProcesses.jl/files/10061313/paper.pdf)

----
- [x] I have read and checked the items on the review checklist.

### CODE REVIEW

File: paper.md [hunk 1]
```diff
@@ -99,6 +99,45 @@ We also implement some features to improve robustness and flexibility of the ens
 
 * The TOML-file interface defined in the \texttt{UQParameters} module allows non-intrusive use of \texttt{EnsembleKalmanProcesses.jl} through \text{TOML} files, which are widely used for configuration files and easily read in any programming language. Given the computer model to calibrate and prior distributions on the parameters, \texttt{EnsembleKalmanProcesses.jl} reads these distributions from a file and, after an iteration of the ensemble Kalman algorithm, writes each member of the updated ensemble to a parameter file. Each of these parameter files can be then read individually to initiate the ensemble of the computer model for the next iteration.
 
+# Pedagogical example
+
+In this example, the complex computer code simulates a sinusoid curve $$f(A,v) = A\sin(\phi + t) + v, \ \forall t \in [0,2\pi],$$ with a random phase shift $\phi$. We assume that we cannot access $f$, but we can access some informative quantities through the map $$G(A,v) = [\max f(A,v) - \min f(A,v), \mathrm{mean} f(A,v)].$$
```
Comment (ilopezgp) [inline paper.md:104]:
```suggestion
In this example, the computer code simulates a sine wave $$f(A,v) = A\sin(t + \varphi) + v, \ \forall t \in [0,2\pi],$$ with a random phase shift $\varphi$. We assume that we cannot evaluate $f$ directly, but we can access some informative quantities through the map $$G(A,v) = [\max f(A,v) - \min f(A,v), \mathrm{mean} f(A,v)].$$
```
This is more standard notation in natural sciences & engineering.

File: paper.md [hunk 2]
```diff
@@ -99,6 +99,45 @@ We also implement some features to improve robustness and flexibility of the ens
...
+
+We are given one sample measurement of $G$, polluted by Gaussian noise $N(0,\Gamma)$, and call this $y$. Our task is to deduce the most likely amplitude $A$ and vertical shift $v$ of the curve that produced the $y$.
```
Comment (ilopezgp) [inline paper.md:106]:
```suggestion
We are given one sample measurement of $G$, polluted by Gaussian noise $N(0,\Gamma)$, and call this $y$. Our task is to deduce the most likely amplitude $A$ and offset $v$ of the curve that produced the $y$. 
```
?

File: paper.md [hunk 3]
```diff
@@ -99,6 +99,45 @@ We also implement some features to improve robustness and flexibility of the ens
...
+We can encode prior information mathematically in the code:
+```julia
+# A is positive, has likely value 2 with standard deviation 1
+# v has likely value 0 with standard deviation 5
+prior_A = constrained_gaussian("amplitude", 2, 1, 0, Inf)
+prior_v = constrained_gaussian("vert_shift", 0, 5, -Inf, Inf)
+prior = combine_distributions([prior_A, prior_v])
+```
+
+To use a basic ensemble method we require one more parameter, the size of the ensemble, which we take to be `N_ensemble = 5`. We now begin solving the problem, by creating the initial ensemble from our prior, and selecting the `Inversion()` tool to perform ensemble Kalman inversion:
```
Comment (ilopezgp) [inline paper.md:118]:
```suggestion
To use a basic ensemble method we need to specify one more hyperparameter, the size of the ensemble, which we take to be `N_ensemble = 5`. We now begin solving the problem, by creating the initial ensemble from our prior, and selecting the `Inversion()` tool to perform ensemble Kalman inversion:
```

File: paper.md [hunk 4]
```diff
@@ -99,6 +99,45 @@ We also implement some features to improve robustness and flexibility of the ens
...
+for i in 1:N_iterations`
+    # get the latest parameter ensemble
+    params_i = get_phi_final(prior, ensemble_kalman_process)
+    # run a simulation for each parameter in the ensemble
+    G_ens = hcat([G(params_i[:, i]) for i in 1:N_ensemble]...)
+    # perform the Kalman update, producing a new ensemble
+    update_ensemble!(ensemble_kalman_process, G_ens)
+end
+```
+The solution of the ensemble Kalman inversion is the final ensemble of parameters. The results are depicted in \autoref{fig:sinusoid}, by evaluating f at these parameters. We see how these methods will calibrate model parameters without estimating the state, as the random shift is not learnt.
```
Comment (ilopezgp) [inline paper.md:136]: _f_ missing math notation

File: paper.md [hunk 5]
```diff
@@ -99,6 +99,45 @@ We also implement some features to improve robustness and flexibility of the ens
...
+
+![Sinusoids produced from initial and final ensembles, and the (un-noised) truth. \label{fig:sinusoid}](sinusoid_output.pdf){width=80%}
```
Comment (ilopezgp) [inline paper.md:138]:
```suggestion
![Sinusoids produced from initial and final ensembles, and the (denoised) truth. \label{fig:sinusoid}](sinusoid_output.pdf){width=80%}
```

File: paper.md (hunk 1 above)
Comment (eviatarbach) [inline paper.md:104]: Why do you specify that we cannot access _f_? If _G_ represents some climate observables, for instance, then _f_ would be the climate model, which we need to be able to run. Maybe I'm misunderstanding what you mean.

File: paper.md (hunk 2 above)
Comment (eviatarbach) [inline paper.md:106]:
```suggestion
We are given one sample measurement of $G$, polluted by Gaussian noise $\mathcal{N}(0,\Gamma)$, and call this $y$. Our task is to deduce the most likely amplitude $A$ and vertical shift $v$ of the curve that produced the $y$. 
```

File: paper.md [hunk 6]
```diff
@@ -99,6 +99,45 @@ We also implement some features to improve robustness and flexibility of the ens
...
 
+# Pedagogical example
+
+In this example, the complex computer code simulates a sinusoid curve $$f(A,v) = A\sin(\phi + t) + v, \ \forall t \in [0,2\pi],$$ with a random phase shift $\phi$. We assume that we cannot access $f$, but we can access some informative quantities through the map $$G(A,v) = [\max f(A,v) - \min f(A,v), \mathrm{mean} f(A,v)].$$
+
+We are given one sample measurement of $G$, polluted by Gaussian noise $N(0,\Gamma)$, and call this $y$. Our task is to deduce the most likely amplitude $A$ and vertical shift $v$ of the curve that produced the $y$. 
+
+
+We can encode prior information mathematically in the code:
```
Comment (eviatarbach) [inline paper.md:109]:
```suggestion
We encode our prior distributions over the parameters:
```

File: paper.md (hunk 4 above)
Comment (eviatarbach) [inline paper.md:136]:
```suggestion
The solution of the ensemble Kalman inversion is the final ensemble of parameters. The results are depicted in \autoref{fig:sinusoid}, by evaluating f at these parameters.

Note that here, the random phase shift does not affect the observables $G(A,v)$, and thus does not affect the results of the calibration. This is a demonstration of a more general principle, that observables can be chosen such that they are not impacted by such "nuisance parameters" which we are not interested in estimating. In the case of climate modelling, for instance, time-averaged observables can be used such that the state becomes unimportant [maybe cite one of the CES papers that discusses this here].
```

Comment (eviatarbach) [reply paper.md:136]: I don't know if this is too much to add, just wanted to make it clearer.

File: paper.md (hunk 5 above)
Comment (eviatarbach) [reply paper.md:138]: "Denoised" suggests that noise was removed. Maybe "non-noisy"?

File: paper.md (hunk 1 above)
Comment (eviatarbach) [reply paper.md:104]:
```suggestion
In this example, the complex computer code simulates a sinusoid curve $$f(A,v) = A\sin(\phi + t) + v, \ \forall t \in [0,2\pi],$$ with a random phase shift $\phi$. We assume that we cannot access $f$, but we can access some informative quantities through the observable map $$G(A,v) = [\max f(A,v) - \min f(A,v), \mathrm{mean} f(A,v)].$$
```

**Review by eviatarbach — CHANGES_REQUESTED** [review]

File: paper.md (hunk 5 above)
Comment (ilopezgp) [reply paper.md:138]: We are denoising it, if we consider the data to be the truth. I guess truth is a tricky word. Do we mean true dynamics = truth, or observation = truth? Typically in ML, people use the latter

File: paper.md (hunk 1 above)
Comment (odunbar) [reply paper.md:104]: done

File: paper.md (hunk 3 above)
Comment (odunbar) [reply paper.md:118]: done

File: paper.md (hunk 2 above)
Comment (odunbar) [reply paper.md:106]: done

File: paper.md (hunk 4 above)
Comment (odunbar) [reply paper.md:136]: I'll add something to this effect when we introduce G. I think the "in the case of climate modeling..." is too much so I've left this out.

File: paper.md (hunk 1 above)
Comment (odunbar) [reply paper.md:104]: I'll have another go at reintroducing f,G

**Comment by eviatarbach** [comment]
Looks good to me!

**Review by eviatarbach — APPROVED** [review]

**Review by ilopezgp — APPROVED** [review]
LGTM

**Comment by ilopezgp** [comment]
Also, do you specify somewhere how you are sampling \varphi in your initial/final ensemble members? Are you drawing it randomly for each member? It seems that most of the variability in the final ensemble actually comes from \varphi and not `A, v`, which is visually misleading.

File: paper.md [hunk 7]
```diff
@@ -99,6 +99,45 @@ We also implement some features to improve robustness and flexibility of the ens
...
+for i in 1:N_iterations`
+    # get the latest parameter ensemble
+    params_i = get_phi_final(prior, ensemble_kalman_process)
+    # run a simulation for each parameter in the ensemble
+    G_ens = hcat([G(params_i[:, i]) for i in 1:N_ensemble]...)
+    # perform the Kalman update, producing a new ensemble
+    update_ensemble!(ensemble_kalman_process, G_ens)
+end
+```
+The solution of the ensemble Kalman inversion is the final ensemble of parameters. The results are depicted in \autoref{fig:sinusoid}, by evaluating $f$ at these parameters. Julia code for this example is provided in the documentation.
```
[the remark this answers is not in the document]
Comment (odunbar) [reply paper.md]: I've added a sentence

**Comment by odunbar** [comment]
https://github.com/openjournals/joss-reviews/issues/4869

**Pull request merged**